hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/AsyncResponsePacketHandler.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.impl.operationservice.impl;
+
+import com.hazelcast.instance.HazelcastThreadGroup;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.Packet;
+import com.hazelcast.spi.impl.PacketHandler;
+import com.hazelcast.spi.impl.operationexecutor.OperationHostileThread;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import static com.hazelcast.instance.OutOfMemoryErrorDispatcher.inspectOutputMemoryError;
+import static com.hazelcast.nio.Packet.HEADER_OP;
+import static com.hazelcast.nio.Packet.HEADER_RESPONSE;
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static com.hazelcast.util.Preconditions.checkTrue;
+
+/**
+ * The AsyncResponsePacketHandler is a PacketHandler that asynchronously process operation-response packets.
+ *
+ * So when a response is received from a remote system, it is put in the workQueue of the ResponseThread.
+ * Then the ResponseThread takes it from this workQueue and calls the {@link PacketHandler} for the
+ * actual processing.
+ *
+ * The reason that the IO thread doesn't immediately deals with the response is that deserializing the
+ * {@link com.hazelcast.spi.impl.operationservice.impl.responses.Response} and let the invocation-future
+ * deal with the response can be rather expensive currently.
+ */
+public class AsyncResponsePacketHandler implements PacketHandler {
+
+    private final ResponseThread responseThread;
+    private final BlockingQueue<Packet> workQueue = new LinkedBlockingQueue<Packet>();
+    private final ILogger logger;
+
+    public AsyncResponsePacketHandler(HazelcastThreadGroup threadGroup,
+                                      ILogger logger,
+                                      PacketHandler responsePacketHandler) {
+        this.logger = logger;
+        this.responseThread = new ResponseThread(threadGroup, responsePacketHandler);
+        responseThread.start();
+    }
+
+    public int getQueueSize() {
+        return workQueue.size();
+    }
+
+    public void shutdown() {
+        responseThread.shutdown();
+    }
+
+    //@Override
```

Comment:
I'll merge that one directly in master when this PR is merged. Else we need to wait again for it to build. Unless of course there is another review comment that needs to get addressed.